subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,286 @@
+		svc->pev[i].event.fd = -1;
+		svc->pev[i].event.events = 0;
+	}
+}
+
+int z_impl_net_socket_service_register(const struct net_socket_service_desc *svc,
+				       struct zsock_pollfd *fds, int len,
+				       void *user_data)
+{
+	int i, ret = -ENOENT;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (!init_done) {
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	if (STRUCT_SECTION_START(net_socket_service_desc) > svc ||
+	    STRUCT_SECTION_END(net_socket_service_desc) <= svc) {
+		goto out;
+	}
+
+	if (fds == NULL) {
+		cleanup_svc_events(svc);
+	} else {
+		if (len > svc->pev_len) {
+			NET_DBG("Too many file descriptors, "
+				"max is %d for service %p",
+				svc->pev_len, svc);
+			ret = -ENOMEM;
+			goto out;
+		}
+
+		for (i = 0; i < len; i++) {
+			svc->pev[i].event = fds[i];
+			svc->pev[i].user_data = user_data;
+		}
+
+		for (i = 0; i < svc->pev_len; i++) {
+			ctx.events[get_idx(svc) + i] = svc->pev[i].event;
+		}
+	}
+
+	/* Tell the thread to re-read the variables */
+	eventfd_write(ctx.events[0].fd, 1);
+	ret = 0;
+
+out:
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static struct net_socket_service_desc *find_svc_and_event(
+	struct zsock_pollfd *pev,
+	struct net_socket_service_event **event)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		for (int i = 0; i < svc->pev_len; i++) {
+			if (svc->pev[i].event.fd == pev->fd) {
+				*event = &svc->pev[i];
+				return svc;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+/* We do not set the user callback to our work struct because we need to
+ * hook into the flow and restore the global poll array so that the next poll
+ * round will not notice it and call the callback again while we are
+ * servicing the callback.
+ */
+void net_socket_service_callback(struct k_work *work)
+{
+	struct net_socket_service_event *pev =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct net_socket_service_desc *svc = pev->svc;
+	struct net_socket_service_event ev = *pev;
+
+	ev.callback(&ev.work);
+
+	/* Copy back the socket fd to the global array because we marked
+	 * it as -1 when triggering the work.
+	 */
+	for (int i = 0; i < svc->pev_len; i++) {
+		ctx.events[get_idx(svc) + i] = svc->pev[i].event;
+	}
+}
+
+static int call_work(struct zsock_pollfd *pev, struct k_work_q *work_q,
+		     struct k_work *work)
+{
+	int ret = 0;
+
+	/* Mark the global fd non pollable so that we do not
+	 * call the callback second time.
+	 */
+	pev->fd = -1;
+
+	if (work->handler == NULL) {
+		/* Synchronous call */
+		net_socket_service_callback(work);
+	} else {
+		if (work_q != NULL) {
+			ret = k_work_submit_to_queue(work_q, work);
+		} else {
+			ret = k_work_submit(work);
+		}
+
+		k_yield();
+	}
+
+	return ret;
+
+}
+
+static int trigger_work(struct zsock_pollfd *pev)
+{
+	struct net_socket_service_event *event;
+	struct net_socket_service_desc *svc;
+
+	svc = find_svc_and_event(pev, &event);
+	if (svc == NULL) {
+		return -ENOENT;
+	}
+
+	event->svc = svc;
+
```

Comment:
Tried this with DHCPv4 server service, generally seems to work, however one thing brought my attention.

In the callback function registered for the service I was checking `event.revents` field of the `struct net_socket_service_event` provided to the callback (to check if there's no POLLERR on a socket, or whether it's really POLLIN that triggered the cb). It seems however that `event->event.revents` is not set here, so the callback could not inspect `revents` as it was always 0.

To me it looks simply like an overlook, but perhaps I miss something, is that intended behavior?

Reply:
> To me it looks simply like an overlook, but perhaps I miss something, is that intended behavior?

No, that was not intended. Will fix.